// _app/widgets/buttons/toggle-button.widget.js
const fs = require('fs');
const domify = require('domify');

class ToggleButton {
    constructor() {
        this.ON_STATE_CLASS = 'on';
        this.OFF_STATE_CLASS = 'off';

        this.buildWidget();
        this.attachEvents();
    }

    buildWidget() {
        const tpl = fs.readFileSync(__dirname + '/toggle-button.widget.html', 'utf8');

        this.elementDom = domify(tpl);

        return this;
    }

    attachEvents() {
        this.elementDom.addEventListener('click', this.toggle.bind(this));
    }

    toggle() {
        if (this.isOn) {
            return this.turnOff();
        }

        return this.turnOn();
    }

    turnOn() {
        this.elementDom.classList.remove(this.OFF_STATE_CLASS);
        this.elementDom.classList.add(this.ON_STATE_CLASS);

        return this;
    }

    turnOff() {
        this.elementDom.classList.remove(this.ON_STATE_CLASS);
        this.elementDom.classList.add(this.OFF_STATE_CLASS);

        return this;
    }

    appendTo(elementSelector) {
        document.querySelector(elementSelector).appendChild(this.elementDom);
    }

    get isOn() {
        return this.elementDom.classList.contains(this.ON_STATE_CLASS);
    }

    get isOff() {
        return ! this.isOn;
    }
}

module.exports = ToggleButton;
